### ISSUE TITLE
Protocols with empty slots rejected from issubclass for being Data protocols

### ISSUE THREAD

**Issue by maxfischer2781** [microsoft/pyright#11025]
**Describe the bug**
A `Protocol` with `__slots__ = ()` is treated as a Data Protocol and rejected from `issubclass` checks. This makes `runtime_checkable` protocols like `AsyncContextManager` unusable in cases that require `issubclass`.

**Code or Screenshots**
A pretty minimal reproducer is this:

```python
from typing import Protocol, runtime_checkable

@runtime_checkable
class Foo(Protocol):
    __slots__ = ()
    def just_some_method(self) -> int: ...

def foo(ac: object):
    return isinstance(ac, type) and issubclass(ac, Foo)
```

The argument `Foo` to `issubclass` is rejected as
> Data protocols (which include non-method attributes) are not allowed in issubclass calls (reportGeneralTypeIssues)

The trigger is the presence of the "attribute" `__slots__ = ()`. I think this is an error, as it actually means the Protocol does *not* have any attributes. At runtime, these protocols are perfectly fine to use in `issubclass`.

**VS Code extension or command-line**
Both CLI and VSC extension report this. Bisecting through the versions, this seems to have been introduced with PyRight v1.1.345 / #6926.

Note in production this only [started failing 4th of October 2025 / pyright 1.1.406](https://github.com/maxfischer2781/asyncstdlib/actions/runs/18242984237) but [still succeeded on 1st of October 2025 / pyright 1.1.405](https://github.com/maxfischer2781/asyncstdlib/actions/runs/18158948804). So in real life code there may be confounding factors to trigger/mask this.

**Comment by erictraut** [comment]
The `Protocol` special form is intended to define structural types. What is your intent in adding a `__slots__` to a structural type definition? Are you doing this so you can explicitly derive from the protocol class. If so, why not define `__slots__` in the derived class?

I think that adding a `__slots__` attribute to a protocol definition is suspect, and I'm not convinced that a type checker should add special casing to allow this.

The typing spec explicitly disallows data protocols to be used in `issubclass` calls. This is done not because the call to `issubclass` will result in a runtime error but because it results in type safety issues.

**Comment by maxfischer2781** [comment]
> What is your intent in adding a `__slots__` to a structural type definition?

This isn't my choice, though I see sense in it. There are quite some of such `Protocol`s in typeshed: 

- [`typing.SupportsXYZ`](https://github.com/python/typeshed/blob/7f2d9484f3694027f8adc962646d0dfb066bf0c9/stdlib/typing.pyi#L464-L508)
- [`contextlib.Abstract(Async)ContextManager`](https://github.com/python/typeshed/blob/7f2d9484f3694027f8adc962646d0dfb066bf0c9/stdlib/contextlib.pyi#L48-L67)

These are unusable for `issubclass` checks at the moment.

> Are you doing this so you can explicitly derive from the protocol class.  If so, why not define `__slots__` in the derived class?

A single baseclass without `__slots__` is sufficient to automatically allow non-slot attributes for all derived classes. It's important that `__slots__ = ()` is present in any baseclass for which the derived classes might want to use `__slots__`.

> I think that adding a `__slots__` attribute to a protocol definition is suspect, and I'm not convinced that a type checker should add special casing to allow this.
>
> The typing spec explicitly disallows data protocols to be used in `issubclass` calls. This is done not because the call to `issubclass` will result in a runtime error but because it results in type safety issues.

This is only about *empty* `__slots__` (i.e. `__slots__ = ()`). The meaning of this is that the protocol itself has explicitly no instance attributes, which should be equivalent to it not being a data protocol.

-----------

Note I expected this to be infectious, but checking indicates that the test seems to be inconsistent. Deriving from such a `Protocol` and adding `__slots__` does not trigger the error – even if the slots are not empty. ([Tested in Playground on v1.1.406](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAK4MYAxmADYA0UIArijFgKYD6pAFs6QNYCGAIwrMaAQQDOcFKQDCYRswAeMALJ8UfNMxABYAFD6AAvUYt2XXoOH7SFPuPFQAIkhDcYACiJgS5CgEoALn0oUKhWVnEKH3EIqABeKA9-ELCAE2ZgKAArOnEYSMg2CGYYDjA0j3FmCmB-KABaAD5MRkCoADouoxMmEvNufiFmGzsHZ20kADdmSokpWXkYJVV1TW0g1NCIqJi4xOStqAys3PzC-pKyiqqausaW1Bh2ro6ehj62TkGrEb1beyOJyTGZpADqSDKohgMBA4g882kcgUyjUGi0IE2ejC4Ui0RgsVYCSSACJxEVWHwYSASTQSeofFwQJTqSSUtj0pkcnkCuTLqVypVqrV6s1Ws9Ot0DHoTlBgGAwB4%2BKR2mABNl3FicUgskhxKh8upSMwlaQaPAEMwtTiwggQE8PHrxHQBACHKaxJIkUsVmj1pj6lAAMTOKl8KB2nxkSiODwAdw4SE4rVsdAyUBQ8gaV0FUCpsKQAjoy3E9T4bgzPjzFGicdmrUwDhdbscpD4NfERxxdodTubY3hypoLjcpBg-kDNqnoRDTjDEeI0YosYTSY4KYoaeYlZQ2YFFTz1MLxeYpbzFcz%2BHbtfrqEbztdA6gbY7Xdt9sYjqbj8BHom9tBCdp2nENMygbRwF0Dkpx7T8%2Bx-d0h3-aZZghKFqVLScZ0rcCQEg-QgA))

```
from typing import Protocol, runtime_checkable, AsyncContextManager

@runtime_checkable
class Direct(Protocol):
    __slots__ = ()
    def just_some_method(self) -> int: ...
@runtime_checkable
class Derived(AsyncContextManager):
    __slots__ = ()
    def just_some_method(self) -> int: ...
@runtime_checkable
class DerivedWithAttrs(AsyncContextManager):
    __slots__ = ("some_attr", "another_attr")
    def just_some_method(self) -> int: ...

def foo(ac: object):
    if isinstance(ac, type):
        print(issubclass(ac, AsyncContextManager))  # Data protocols (which include non-method attributes) are not allowed in issubclass calls
        print(issubclass(ac, Direct))               # Data protocols (which include non-method attributes) are not allowed in issubclass calls
        print(issubclass(ac, Derived))              # no error
        print(issubclass(ac, DerivedWithAttrs))     # no error
```
Perhaps this should be fixed as well, one way or another.

**Comment by erictraut** [comment]
I think pyright is following the typing spec in this matter, so I'm going to close the issue as "as designed". Refer to [this section](https://typing.python.org/en/latest/spec/protocol.html#runtime-checkable-decorator-and-narrowing-types-by-isinstance) for details.

The typing spec could be amended to special-case `__slots__`, but this would require community discussion and consensus. For details on how to amend the typing spec, refer to [this documentation](https://github.com/python/typing-council#decisions). If you'd like to pursue this, a good starting point is a new thread on discourse in the [typing forum](https://discuss.python.org/c/typing/32).
